src/main/java/com/synacy/gradprogram/spock/exercise/RefundService.java

```diff
@@ -4,14 +4,36 @@
 
 public class RefundService {
 
-  public BigDecimal calculateRefund() {
-    // TODO: Implement me. Full refund if cancel reason is due to damaged item.
-    //  Also full refund if the order was cancelled within 3 days of order date, else refund half of the total cost.
-    return null;
+  private final DateUtils dateUtils;
+  private final RefundRepository refundRepository;
+  private final OrderRepository orderRepository;
+
+  public RefundService(DateUtils dateUtils, RefundRepository refundRepository, OrderRepository orderRepository) {
+    this.dateUtils = dateUtils;
+    this.refundRepository = refundRepository;
+    this.orderRepository = orderRepository;
   }
 
-  private void createAndSaveRefundRequest() {
-    // TODO: Implement me. Creates a TO_PROCESS refund request and saves it to the database
+  private BigDecimal calculateRefund(Order order, CancelOrderRequest cancelOrderRequest) {
+    final double DIVISOR = 2.0;
+
+    if (dateUtils.isDateWithinRefundPeriod(order.getDateOrdered()) || cancelOrderRequest.getReason() == CancelReason.DAMAGED) {
+      return BigDecimal.valueOf(order.getTotalCost());
+    }
+
+    return BigDecimal.valueOf(order.getTotalCost() / DIVISOR);
+  }
+
+  public void createAndSaveRefundRequest(CancelOrderRequest cancelOrderRequest) {
+    Order order = orderRepository.fetchOrderById(cancelOrderRequest.getOrderId()).get();
```

Comment:
We already fetched the Order in the `cancelOrder`, we can just include as another input instead :+1: